src/core_plugins/kibana/public/dashboard/actions/embeddables.ts

```diff
@@ -0,0 +1,140 @@
+import { PanelId, PanelState } from '../selectors';
+import { updatePanel } from './panels';
+
+import { EmbeddableMetadata, EmbeddableState } from 'ui/embeddable';
+import {
+  getEmbeddableCustomization,
+  getPanel,
+} from '../../selectors/dashboard_selectors';
+
+export type EmbeddableIsInitializingActionPayload = string;
+
+export enum EmbeddableActionTypeKeys {
+  EMBEDDABLE_IS_INITIALIZING = 'EMBEDDABLE_IS_INITIALIZING',
+  EMBEDDABLE_IS_INITIALIZED = 'EMBEDDABLE_IS_INITIALIZED',
+  SET_STAGED_FILTER = 'SET_STAGED_FILTER',
+  CLEAR_STAGED_FILTERS = 'CLEAR_STAGED_FILTERS',
+  EMBEDDABLE_ERROR = 'EMBEDDABLE_ERROR',
+}
+
+export interface EmbeddableIsInitializingAction extends Action {
+  type: EmbeddableActionTypeKeys.EMBEDDABLE_IS_INITIALIZING;
+  payload: EmbeddableIsInitializingActionPayload;
+}
+
+export interface EmbeddableIsInitializedActionPayload {
+  panelId: PanelId;
+  metadata: EmbeddableMetadata;
+}
+
+export interface EmbeddableIsInitializedAction extends Action {
+  type: EmbeddableActionTypeKeys.EMBEDDABLE_IS_INITIALIZED;
+  payload: EmbeddableIsInitializedActionPayload;
+}
+
+export interface SetStagedFilterActionPayload {
+  panelId: PanelId;
+  stagedFilter: object;
+}
+
+export interface SetStagedFilterAction extends Action {
+  type: EmbeddableActionTypeKeys.SET_STAGED_FILTER;
+  payload: SetStagedFilterActionPayload;
+}
+
+export interface ClearStagedFiltersAction extends Action {
+  type: EmbeddableActionTypeKeys.CLEAR_STAGED_FILTERS;
+  payload: undefined;
+}
+
+export interface EmbeddableErrorActionPayload {
+  error: string | object;
+  panelId: PanelId;
+}
+
+export interface EmbeddableErrorAction extends Action {
+  type: EmbeddableActionTypeKeys.EMBEDDABLE_ERROR;
+  payload: EmbeddableErrorActionPayload;
+}
+
+export type EmbeddableActions =
+  | EmbeddableIsInitializingAction
+  | EmbeddableIsInitializedAction
+  | ClearStagedFiltersAction
+  | SetStagedFilterAction
+  | EmbeddableErrorAction;
+
+export const embeddableIsInitializing = createAction<
+  EmbeddableIsInitializingActionPayload
+>(EmbeddableActionTypeKeys.EMBEDDABLE_IS_INITIALIZING);
+export const embeddableIsInitialized = createAction<
+  EmbeddableIsInitializedActionPayload
+>(EmbeddableActionTypeKeys.EMBEDDABLE_IS_INITIALIZED);
+export const setStagedFilter = createAction<SetStagedFilterActionPayload>(
+  EmbeddableActionTypeKeys.SET_STAGED_FILTER
+);
+export const clearStagedFilters = createAction<void>(
+  EmbeddableActionTypeKeys.CLEAR_STAGED_FILTERS
+);
+export const embeddableError = createAction<EmbeddableErrorActionPayload>(
+  EmbeddableActionTypeKeys.EMBEDDABLE_ERROR
+);
+
+/**
+ * The main point of communication from the embeddable to the dashboard. Any time state in the embeddable
+ * changes, this function will be called. The data is then extracted from EmbeddableState and stored in
+ * redux so the appropriate actions are taken and UI updated.
+ *
+ * @param {string} panelId - the id of the panel whose state has changed.
```

Comment:
Or rather remove the type at all, since we are using typescript and jsdoc types are redundant.

Reply:
> Or rather remove the type at all, since we are using typescript and jsdoc types are redundant.

That's something that everyone in Kibana Team suggests, it seems, so feel free to make it a rule :)

I'm a little skeptical though as I had a chance to transpile entire new platform TS to JS at some point :see_no_evil: (see [this thread](https://github.com/elastic/kibana/pull/18562#discussion_r190148027) if you're curious where my skepticism comes from).

Reply:
> I also think there are a bunch of tools that can convert proper JSDoc types to say Flow annotations or something like this if we change our mind at some point.

It'd be cool if there was also a tool to sync TS types -> jsDoc types (maybe there is... I haven't actually looked). Then it would be a single command and we could have both, rather than manually keeping them in sync, which is prone to error (like here).  For now I'll just leave as is and wait for the broader conversation about this.  :)

Reply:
:+1: